src/prefixtrie.cpp

```diff
@@ -0,0 +1,486 @@
+{
+    if (key.empty()) return it;
+    auto shared = it.node.lock();
+    assert(shared);
+    return find(key, shared, end());
+}
+
+template <typename TKey, typename TData>
+TData& CPrefixTrie<TKey, TData>::at(const TKey& key)
+{
+    return find(key).data();
+}
+
+template <typename TKey, typename TData>
+std::vector<typename CPrefixTrie<TKey, TData>::iterator> CPrefixTrie<TKey, TData>::nodes(const TKey& key) const
+{
+    std::vector<iterator> ret;
+    if (empty()) return ret;
+    ret.reserve(1 + key.size());
+    ret.emplace_back(TKey{}, root);
+    if (key.empty()) return ret;
+    TKey name;
+    using CBType = callback<std::shared_ptr<Node>>;
+    CBType cb = [&name, &ret](const TKey& key, std::shared_ptr<Node> node) {
+        name.insert(name.end(), key.begin(), key.end());
+        ret.emplace_back(name, node);
+    };
+    find(key, root, cb);
+    return ret;
+}
+
+template <typename TKey, typename TData>
+bool CPrefixTrie<TKey, TData>::erase(const TKey& key)
+{
+    auto size_was = height();
+    if (key.empty()) {
+        root->data = std::make_shared<TData>();
+    } else {
+        erase(key, root);
+    }
+    return size_was != height();
+}
+
+template <typename TKey, typename TData>
+void CPrefixTrie<TKey, TData>::clear()
+{
+    size = 0;
+    root->data = std::make_shared<TData>();
+    root->children.clear();
+}
+
+template <typename TKey, typename TData>
+bool CPrefixTrie<TKey, TData>::empty() const
+{
+    return height() == 0;
+}
+
+template <typename TKey, typename TData>
+std::size_t CPrefixTrie<TKey, TData>::height() const
+{
+    return size + (root->data->empty() ? 0 : 1);
+}
+
+template <typename TKey, typename TData>
+typename CPrefixTrie<TKey, TData>::iterator CPrefixTrie<TKey, TData>::begin()
+{
+    return find(TKey());
+}
+
+template <typename TKey, typename TData>
+typename CPrefixTrie<TKey, TData>::iterator CPrefixTrie<TKey, TData>::end()
+{
+    return iterator{TKey(), std::shared_ptr<Node>{}};
+}
+
+template <typename TKey, typename TData>
+typename CPrefixTrie<TKey, TData>::const_iterator CPrefixTrie<TKey, TData>::cbegin()
+{
+    return find(TKey());
+}
+
+template <typename TKey, typename TData>
+typename CPrefixTrie<TKey, TData>::const_iterator CPrefixTrie<TKey, TData>::cend()
+{
+    return const_iterator{TKey(), std::shared_ptr<Node>{}};
+}
+
+template <typename TKey, typename TData>
+typename CPrefixTrie<TKey, TData>::const_iterator CPrefixTrie<TKey, TData>::begin() const
+{
+    return find(TKey());
+}
+
+template <typename TKey, typename TData>
+typename CPrefixTrie<TKey, TData>::const_iterator CPrefixTrie<TKey, TData>::end() const
+{
+    return const_iterator{TKey(), std::shared_ptr<Node>{}};
+}
+
+using Key = std::string;
+using Data = CClaimTrieData;
+using Trie = CPrefixTrie<Key, Data>;
+using iterator = Trie::iterator;
+using const_iterator = Trie::const_iterator;
+
+template class CPrefixTrie<Key, Data>;
```

Comment:
This and the lines below it are a c++ nuance. You have to instantiate a specific type of a template or you need the full definition in the header file. As a native C# programmer, I prefer the latter. @bvbfan prefers the former (as it is here).

Reply:
In some project is used *.impl file ext (not *.cpp) and gets include where it's needed, i'm ok with that too. But all in header makes things to look redundant and hard to read.